recipes-core/images/includes/nilrt-image-base.inc

```diff
@@ -6,6 +6,16 @@ do_rootfs[depends] += "shadow-native:do_populate_sysroot"
 # kernel recipe requires depmodwrapper to populate modules.dep
 do_rootfs[depends] += "depmodwrapper-cross:do_populate_sysroot"
 
+# Build package index before trying to build rootfs
+do_rootfs[depends] += "${@oe.utils.conditional('BUILD_IMAGES_FROM_FEEDS', '1', 'package-index:do_package_index', '', d)}"
+do_package_index[nostamp] = "1"
+do_package_index[depends] += "${PACKAGEINDEXDEPS}"
```

Comment:
I heavily debated making this part of `rootfs_<packagetype>.bbclass` instead, but there are a couple of things I think would preclude that:
- `package-index` should probably be a class instead of a task defined in a standalone recipe.
- Building package indexes only impacts local feeds. It was unclear to me if there needed to be differentiation to skip the step if none of the feeds for building were local.

Reply:
I think the only reason we wouldn't want to add this task-dependency to all images when doing `BUILD_FROM_FEEDS` is that it would be unnecessary for all images beyond the first - and it is kind of slow.

>package-index should probably be a class instead of a task defined in a standalone recipe.

I like the existing implementation as an independent target, since it enables us to build indexes for the feeds in the OE_feeds component without invoking an image target.

>Building package indexes only impacts local feeds. It was unclear to me if there needed to be differentiation to skip the step if none of the feeds for building were local.

Nah; doesn't hurt to run the task if there are no local feeds. Not worth the complexity of adding the skip-logic, I don't think.

Reply:
> I think the only reason we wouldn't want to add this task-dependency to all images when doing `BUILD_FROM_FEEDS` is that it would be unnecessary for all images beyond the first - and it is kind of slow.

It actually seems to run pretty quickly for me as part of the build. Any suggestions on how to benchmark the time it adds to a build?

> I like the existing implementation as an independent target, since it enables us to build indexes for the feeds in the OE_feeds component without invoking an image target.

We can actually have our cake and eat it too. Example case: `meta-toolchain` literally just [inherits](https://github.com/ni/openembedded-core/blob/nilrt/master/hardknott/meta/recipes-core/meta/meta-toolchain.bb#LL15C1-L15C21) `populate-sdk` and builds that. My plan was to keep the specific `package-index` recipe but move the logic to a class instead of having it live in the recipe.

> Nah; doesn't hurt to run the task if there are no local feeds. Not worth the complexity of adding the skip-logic, I don't think.

Cool, agreed then.

Reply:
> I think the only reason we wouldn't want to add this task-dependency to all images when doing BUILD_FROM_FEEDS is that it would be unnecessary for all images beyond the first - and it is kind of slow.

It used to take ~10 mins previously but we [disabled](https://github.com/ni/openembedded-core/pull/70) building package filelist generation and I believe it takes ~10secs now.

I'm not opposed to adding indexing stage into one of the existing recipes but I'd prefer it go in packagegroup recipes instead of image recipes. Better yet, make it part of [packagegroup.bbclass](https://github.com/ni/openembedded-core/blob/nilrt/master/kirkstone/meta/classes/packagegroup.bbclass#L1).

Reply:
> I'm not opposed to adding indexing stage into one of the existing recipes but I'd prefer it go in packagegroup recipes instead of image recipes. Better yet, make it part of [packagegroup.bbclass](https://github.com/ni/openembedded-core/blob/nilrt/master/kirkstone/meta/classes/packagegroup.bbclass#L1).

I prefer having it be a prerequisite step for `do_rootfs` instead for a couple of reasons:
- The package index isn't specifically required when building a packagegroup, but it _is_ required to build an image from feeds. Additionally, I was gating it logically on BIFF being enabled.
- If the packagegroup is where the index build occurs, I'm not sure we get to keep the one command image build that is a big advantage of this to me. E.g., you would have to build the packagegroup and then the image, whereas having it build the index as part of the image build results in only having to build the image.
   - As an aside, I'm actually a bit confused on this. Should BIFF result in all dependencies being built before trying to build the image? If not, then maybe adding it to the packagegroup is right. Or even just having the class and inheriting it in our specific `packagefeed` packagegroups since packagegroups aren't necessarily feeds.